zerver/views/streams.py

```diff
@@ -75,12 +77,69 @@
 from zerver.lib.typed_endpoint import ApiParamConfig, PathOnly, typed_endpoint
 from zerver.lib.typed_endpoint_validators import check_color, check_int_in_validator
 from zerver.lib.user_groups import access_user_group_for_setting
-from zerver.lib.users import access_user_by_email, access_user_by_id
+from zerver.lib.users import access_user_by_email, access_user_by_id, access_user_common
 from zerver.lib.utils import assert_is_not_none
 from zerver.models import NamedUserGroup, Realm, Stream, UserProfile
 from zerver.models.users import get_system_bot
 
 
+def bulk_principals_to_user_profiles(
+    agent: UserProfile, principals: list[str] | list[int]
+) -> set[UserProfile]:
+    # principals is guaranteed to be non-empty and to have the same type of elements, int or str,
+    # so the following check is safe and enough.
+
+    # principlas are user emails.
+    if isinstance(principals[0], str):
```

Comment:
Could this be spelled `isinstance(principals, list[str])`? That might let you delete your comment about the type; it's written right there in the type annotation.

Reply:
I think it'd also avoid the `type: ignore` issues below, because mypy would narrow the type properly here.

Reply:
> Could this be spelled `isinstance(principals, list[str])`?

Unfortunately no, raises `TypeError: 'type' object is not subscriptable`

Reply:
I wonder if `List[str]` works for that.

Reply:
Nope, `TypeError: Subscripted generics cannot be used with class and instance checks`

Reply:
By simple deduction, if `principals` is only `str`s OR only `int`s (because if not, will raise an error before executing even executing anything the endpoint), then checking the first element only is more than enough.
[see my test in CZO](https://chat.zulip.org/#narrow/stream/49-development-help/topic/stream.20principals/near/1909001)

Reply:
Yeah, logically it's enough, it's just annoying that mypy seems to not be inferring the overall type.